### PR TITLE
Implement Legendary Heartbeat

### PR DESCRIPTION
**Pull request by flagrama**
Only the Beedrill line so far, but figure it best to put my progress up as I complete it. Probably easier to review each commit as it comes.

### CODE REVIEW

**Comment by flagrama** [comment]
[G] Implemented

**Comment by flagrama** [comment]
[Legendary Heartbeat](https://trello.com/c/11hqxKLd/835-legendary-heartbeat)

**Comment by flagrama** [comment]
[L] and [P] Implemented. Added a Metronome static for Clefairy to use, will adapt and implement for other cards in a separate PR.

**Comment by flagrama** [comment]
All Pokémon are now implemented. Just Trainers, Special Energy, and reprints remaining.

**Comment by flagrama** [comment]
Should be ready!

**Comment by flagrama** [comment]
Can probably remove from Collections and just get a Vivid Voltage groovy set up copying from here and Amazing Volt Tackle where appropriate for testing since it's now easily available: https://www.pokebeach.com/2020/10/vivid-voltage-english-set-list

File: src/tcgwars/logic/impl/gen8/LegendaryHeartbeat.groovy [hunk 1]
```diff
@@ -1198,11 +1274,23 @@ public enum LegendaryHeartbeat implements LogicCardInfo {
...
+          attackRequirement {
+            assert self.energyCards.any { it.cardTypes.contains BASIC_ENERGY } : "No Basic Energy attached to $self"
+          }
           onAttack {
-            damage 80
+            Set<Type> energyTypes = []
+            CardList discardEnergies = []
+            self.energyCards.each {
+              if (!it.cardTypes.contains(BASIC_ENERGY)) return
+              energyTypes.add it.energyTypes.get(0)[0]
```
Comment (axpendix) [inline src/tcgwars/logic/impl/gen8/LegendaryHeartbeat.groovy:1287]:
```suggestion
              energyTypes.add it.basicType
```

File: src/tcgwars/logic/impl/gen8/LegendaryHeartbeat.groovy [hunk 2]
```diff
@@ -1213,15 +1301,19 @@ public enum LegendaryHeartbeat implements LogicCardInfo {
         move "Minor ErrandRunning", {
           text " Search your deck for up to 2 basic Energy cards, reveal them, and put them into your hand. Then, shuffle your deck."
           energyCost C
-          attackRequirement {}
+          attackRequirement {
+            assert my.deck : "Your deck is empty"
+          }
           onAttack {
-
+            deck.search(max:2,{ it.cardTypes.contiains BASIC_ENERGY })
```
Comment (axpendix) [inline src/tcgwars/logic/impl/gen8/LegendaryHeartbeat.groovy:1308]:
```suggestion
            deck.search(max:2, cardTypeFilter(BASIC_ENERGY))
```

**Review by axpendix — COMMENTED** [review]
Hi Vincent, I skimmed through the changes and added some comments, not sure I could cover them all but think merging would be good after comments are resolved. Thanks!

File: src/tcgwars/logic/impl/gen8/LegendaryHeartbeat.groovy [hunk 3]
```diff
@@ -605,17 +679,20 @@ public enum LegendaryHeartbeat implements LogicCardInfo {
...
         move "Tail Order", {
           text "30 damage. You may have your opponent shuffle their hand into their deck and draw 4 cards."
           energyCost P, C
-          attackRequirement {}
           onAttack {
             damage 30
+            if (confirm("Have your opponent shuffle their hand into their deck and draw 4 cards?")) {
+              opp.hand.moveTo hidden:true, opp.deck
+              shuffleDeck null, TargetPlayer.OPPONENT
+              draw 4
```
[the remark this answers is not in the document]
Comment (flagrama) [reply src/tcgwars/logic/impl/gen8/LegendaryHeartbeat.groovy:694]:
```suggestion
              draw 4, TargetPlayer.OPPONENT
```

File: src/tcgwars/logic/impl/gen8/LegendaryHeartbeat.groovy [hunk 4]
```diff
@@ -785,17 +842,20 @@ public enum LegendaryHeartbeat implements LogicCardInfo {
...
           energyCost F, C, C
-          attackRequirement {}
+          attackRequirement {
+            def prizes = my.prizeCardSet.size()
+            assert prizes != 2 && prizes != 4 && prizes != 6 : "You have exactly 2, 4, or 6 Prize cards remaining"
+          }
           onAttack {
-            damage 180
+            def prizes = my.prizeCardSet.size()
+            if (prizes != 2 && prizes != 4 && prizes != 6) damage 180
```
[the remark this answers is not in the document]
Comment (flagrama) [reply src/tcgwars/logic/impl/gen8/LegendaryHeartbeat.groovy:858]: Even if an attack is copied via Metronome or other attack copying moves? Mt Gufo mentioned that as a concern in the past.

[the remark this answers is not in the document]
Comment (flagrama) [reply src/tcgwars/logic/impl/gen8/LegendaryHeartbeat.groovy:857]: Just tested it. Looks like the requirements do prevent in the case of base set Clefairy's Metronome so I know this is correct. I'll have to test it with other move copying effects and the static I created to make sure it properly works in those cases too.

**Pull request merged**